// @lingui/cli/api/formats/lingui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _toConsumableArray2 = require("babel-runtime/helpers/toConsumableArray");

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _defineProperty2 = require("babel-runtime/helpers/defineProperty");

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _stringify = require("babel-runtime/core-js/json/stringify");

var _stringify2 = _interopRequireDefault(_stringify);

var _fs = require("fs");

var _fs2 = _interopRequireDefault(_fs);

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _mkdirp = require("mkdirp");

var _mkdirp2 = _interopRequireDefault(_mkdirp);

var _glob = require("glob");

var _glob2 = _interopRequireDefault(_glob);

var _ramda = require("ramda");

var _ramda2 = _interopRequireDefault(_ramda);

var _locales = require("./utils/locales");

var locales = _interopRequireWildcard(_locales);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sourceFilename = _path2.default.join("{locale}", "messages.json");
var compiledFilename = _path2.default.join("{locale}", "messages.js");

exports.default = function (config) {
  return {
    formatFilename: function formatFilename(pattern, locale) {
      return pattern.replace("{locale}", locale);
    },
    write: function write(locale, messages) {
      var filename = _path2.default.join(config.localeDir, this.formatFilename(sourceFilename, locale));

      var created = !_fs2.default.existsSync(filename);
      _fs2.default.writeFileSync(filename, (0, _stringify2.default)(messages, null, 2));
      return [created, filename];
    },
    read: function read(locale) {
      var filename = _path2.default.join(config.localeDir, this.formatFilename(sourceFilename, locale));

      if (!_fs2.default.existsSync(filename)) return null;

      var raw = _fs2.default.readFileSync(filename).toString();

      try {
        return JSON.parse(raw);
      } catch (e) {
        return null;
      }
    },
    merge: function merge(nextCatalog) {
      var _this = this;

      var nextKeys = _ramda2.default.keys(nextCatalog);

      return _ramda2.default.mergeAll(this.getLocales().map(function (locale) {
        var prevCatalog = _this.read(locale);
        if (!prevCatalog) return nextCatalog;

        var prevKeys = _ramda2.default.keys(prevCatalog);

        var newKeys = _ramda2.default.difference(nextKeys, prevKeys);
        var mergeKeys = _ramda2.default.intersection(nextKeys, prevKeys);
        var obsoleteKeys = _ramda2.default.difference(prevKeys, nextKeys);

        // Initialize new catalog with new keys
        var newMessages = _ramda2.default.mapObjIndexed(function (message, key) {
          return (0, _extends3.default)({
            translation: config.sourceLocale === locale ? key : ""
          }, message);
        }, _ramda2.default.pick(newKeys, nextCatalog));

        // Merge translations from previous catalog
        var mergedMessages = mergeKeys.map(function (key) {
          return (0, _defineProperty3.default)({}, key, (0, _extends3.default)({
            translation: prevCatalog[key].translation
          }, _ramda2.default.omit(["obsolete, translation"], nextCatalog[key])));
        });

        // Mark all remaining translations as obsolete
        var obsoleteMessages = obsoleteKeys.map(function (key) {
          return (0, _defineProperty3.default)({}, key, (0, _extends3.default)({}, prevCatalog[key], {
            obsolete: true
          }));
        });

        var newCatalog = _ramda2.default.mergeAll([newMessages].concat((0, _toConsumableArray3.default)(mergedMessages), (0, _toConsumableArray3.default)(obsoleteMessages)));
        return (0, _defineProperty3.default)({}, locale, newCatalog);
      }));
    },
    getTranslation: function getTranslation(catalogs, locale, key, _ref4) {
      var fallbackLocale = _ref4.fallbackLocale,
          sourceLocale = _ref4.sourceLocale;

      var getTranslation = function getTranslation(locale) {
        return catalogs[locale][key].translation;
      };

      return (
        // Get translation in target locale
        getTranslation(locale) ||
        // Get translation in fallbackLocale (if any)
        fallbackLocale && getTranslation(fallbackLocale) ||
        // Get message default
        catalogs[locale][key].defaults ||
        // If sourceLocale is either target locale of fallback one, use key
        sourceLocale && sourceLocale === locale && key || sourceLocale && fallbackLocale && sourceLocale === fallbackLocale && key ||
        // Otherwise no translation is available
        undefined
      );
    },
    writeCompiled: function writeCompiled(locale, content) {
      var filename = _path2.default.join(config.localeDir, this.formatFilename(compiledFilename, locale));

      _fs2.default.writeFileSync(filename, content);
      return filename;
    },
    getLocale: function getLocale(filename) {
      var filenameRe = new RegExp(this.formatFilename(sourceFilename, locales.localeRe.source));
      var match = filenameRe.exec(filename);
      if (!match) return null;

      return match[1];
    },
    getLocales: function getLocales() {
      var _this2 = this;

      var pattern = _path2.default.join(config.localeDir, this.formatFilename(sourceFilename, "*"));

      return _glob2.default.sync(pattern).map(function (filename) {
        return _this2.getLocale(filename);
      }).filter(Boolean);
    },
    addLocale: function addLocale(locale) {
      if (!locales.isValid(locale)) {
        return [false, null];
      }

      var filename = _path2.default.join(config.localeDir, this.formatFilename(sourceFilename, locale));

      if (!_fs2.default.existsSync(filename)) {
        var dirname = _path2.default.dirname(filename);

        _mkdirp2.default.sync(dirname);
        this.write(locale, {});

        return [true, filename];
      }

      return [false, filename];
    }
  };
};
