// @lingui/cli/api/formats/minimal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _ramda = require("ramda");

var _ramda2 = _interopRequireDefault(_ramda);

var _lingui = require("./lingui");

var _lingui2 = _interopRequireDefault(_lingui);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function deserialize(catalog) {
  return _ramda2.default.map(function (translation) {
    return {
      translation: translation,
      defaults: "",
      origin: []
    };
  }, catalog);
}

function serialize(catalog) {
  return _ramda2.default.map(function (message) {
    return message.translation || "";
  }, catalog);
}

exports.default = function (config) {
  var linguiFormat = (0, _lingui2.default)(config);

  return (0, _extends3.default)({}, linguiFormat, {
    write: function write(locale, messages) {
      return linguiFormat.write(locale, serialize(messages));
    },
    read: function read(locale) {
      return deserialize(linguiFormat.read(locale));
    }
  });
};
